In [1]:
# import librairies 
import pandas as pd 
import numpy as np 
import os 

In [33]:
#Creating a path 
path = r'C:\Users\fairo\03-2023 Instacart Basket Analysis'

In [42]:
# import data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
# Create a dataframe
df_test = pd.DataFrame()

In [5]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [6]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [7]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [8]:
df_test['mix'] = df_test['mix'].astype('str')

In [9]:
# Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [11]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
df_prods.shape

(49693, 5)

In [13]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [14]:
df_prods_clean.shape

(49677, 5)

In [15]:
# Duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [16]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
df_prods_clean.shape

(49677, 5)

In [18]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [19]:
df_prods_clean_no_dups.shape

(49672, 5)

In [39]:
# exporting df_prods
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task

In [43]:
2.# data consistency Checks 
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In the order_dow column there is consistency because the minimum is 0 and the maximum 6 which is consistent with the number of days in a week. The 50th percentile also is 3 which is the correct value. 
There is no negative value either. 
In conclusion there is nothing that look off in this.

In [46]:
3.# Checking for mixed-type data
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

No mixed data was found.

In [50]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB


In [51]:
5.# Checking for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There is 206209 missing values in days-since-prior-order, which can be explained by the fact that these customers have not ordered before so we can say that they are first time customers. 

In [52]:
6.# Addressing the missing values
df_first_time_cust = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [53]:
df_first_time_cust

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

I created a new variable that acts like a flag based on the missing value. Which will be a frist time customer variable.

In [55]:
7.# Checking for duplicates 
df_dups2 = df_ords[df_ords.duplicated()]

In [56]:
df_dups2

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

No duplicates were found

In [58]:
9.# Exporting data
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))